[JIT] Add support for default argument values to Torchbind

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51253 [JIT] Add support for default argument values to Torchbind**

**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

Commit:
[JIT] Add support for default argument values to Torchbind

**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

[ghstack-poisoned]

tugsbayasgalan commented on test/jit/test_torchbind.py:
```diff
@@ -349,3 +349,26 @@ def test_lambda_as_constructor(self):
+
+        self.checkScript(fn, ())
+
+        def gn() -> int:
```
should we also test this case outside of torchscript?

SplitInfinity replied:
`checkScript` tests that, right?

tugsbayasgalan replied:
oh right, my bad

tugsbayasgalan commented on torch/custom_class.h:
```diff
@@ -18,6 +18,32 @@
+/// when registering methods for custom classes.
+///     static auto register_foo = torch::class_<Foo>("myclasses", "Foo")
+///       .def("myMethod", &Foo::myMethod, {torch::arg("name") = name});
+struct arg {
```
this is pretty cool

SplitInfinity replied:
I think the only difference between this and pybind is that in pybind you don't need to pass a list; they use variadic templates to take an arbitrary number of arguments. We *could* do that but it would make the code a little more complex.

gmagogsfm replied:
Could we use variadic arguments as well? Though this is only a minor friction when users migrate from pybind to torchbind, but less friction is still better.

SplitInfinity replied:
I updated the PR to use variadic templates to allow a more pybind-like API but I am not confident in my C++ metaprogramming.

SplitInfinity replied:
It seems we are running into CI errors due to differences between gcc and clang and how they handle parameter packs: https://stackoverflow.com/questions/29098835/can-parameter-pack-function-arguments-be-defaulted

tugsbayasgalan reviewed (approved):
LGTM

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

[ghstack-poisoned]

gmagogsfm commented on torch/custom_class.h:
```diff
@@ -18,6 +18,32 @@
+  // The name of the argument. This is copied to the schema; argument
+  // names cannot be extracted from the C++ declaration.
+  std::string name_;
+  // IValue's default constructor makes it None, which not distinguishable from
```
s/which not/which is not/

Could we have a torch::arg::None method that boils down to an empty IValue? This would make the API more intuitive and resembles Pybind more.

gmagogsfm commented on torch/custom_class.h:
```diff
@@ -98,23 +124,28 @@ class class_ {
+      detail::types<void, Types...>,
+      std::string doc_string = "",
+      std::initializer_list<arg> default_args =
+          {}) { // Used in combination with
```
The format here looks pretty weird. Did clang-format got fooled by the line break in comment?

gmagogsfm commented on torch/custom_class.h:
```diff
@@ -263,11 +298,46 @@ class class_ {
+      std::string doc_string = "",
+      std::initializer_list<arg> default_args = {}) {
     auto qualMethodName = qualClassName + "." + name;
     auto schema = c10::inferFunctionSchemaSingleReturn<Func>(std::move(name), "");
```
I think `inferFunctionSchemaSingleReturn` is only used by TorchBind, could we merge all the logic below related to schema construction into `inferFunctionSchemaSingleReturn`?

SplitInfinity replied:
I would prefer to keep schema inference + attaching default args separatel.

gmagogsfm commented on torch/custom_class.h:
```diff
@@ -263,11 +298,46 @@ class class_ {
+      std::string name,
+      Func func,
+      std::string doc_string = "",
+      std::initializer_list<arg> default_args = {}) {
```
Nit:

Could we consider changing the term `default argument` to `keyword argument` or simply `arg` throughout the PR? the structure `arg` isn't really about default values, they describe all aspects of an argument. IMHO, It would be much clearer.

SplitInfinity replied:
https://pybind11.readthedocs.io/en/stable/basics.html#default-args

The term "default argument" is used here in the pybind documentation

gmagogsfm replied:
`py::arg` is similarly used in more than just expressing default argument, see here:
https://pybind11.readthedocs.io/en/stable/basics.html#keyword-arguments

But this is really minor, up to you.

gmagogsfm commented on torch/custom_class.h:
```diff
@@ -263,11 +298,46 @@ class class_ {
+
+      new_args.emplace_back(old_args[0]);
+      for (size_t i = 0; i < default_args_v.size(); ++i) {
+        auto& arg = old_args[i+1];
```
Nit:

Add comment "skips self"

gmagogsfm reviewed (commented):
Nice! just some minor suggestions.

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

[ghstack-poisoned]

SplitInfinity commented:
@gmagogsfm 

I updated the PR to provide an API closer to pybind but now the clang builds are failing due to a difference between how gcc and clang handle parameter packs: https://stackoverflow.com/questions/29098835/can-parameter-pack-function-arguments-be-defaulted

SplitInfinity commented:
Update: it is probably not possible to use parameter packs like this to accept a variable number of `torch::arg` arguments like pybind does with `py::arg` unless we upgrade to clang11: https://reviews.llvm.org/rGb5f2c4e45b8d54063051e6955cef0bbb7b6ab0f8

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

[ghstack-poisoned]

gmagogsfm commented on torch/custom_class.h:
```diff
@@ -18,6 +18,41 @@
+  // IValue's default constructor makes it None, which is not distinguishable from
+  // an actual, user-provided default value that is None. This boolean
+  // helps distinguish between the two cases.
+  bool value_set_;
```
Optional:

Seems we can use c10::Optional to replace this pair of <bool, IValue> and make code cleaner.

SplitInfinity replied:
NOT Optional! This is a great suggestion.

gmagogsfm reviewed (approved):
Looks good, thanks for implementing it!

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for default argument values to Torchbind"


**Summary**
This commit adds support to Torchbind for specifying default values for
arguments of custom class methods.

**Test Plan**
This commit adds a unit test to `test_torchbind.py` that exercises this
feature.

Differential Revision: [D26131529](https://our.internmc.facebook.com/intern/diff/D26131529)

[ghstack-poisoned]
